### ISSUE TITLE
Properties text not editable (umlet Eclipse plugin 14.2, Mac OS X, Eclipse Neon)

### ISSUE THREAD

**Issue by dimitrijer** [umlet/umlet#399]
First of all, I really dig this, it's simple, small and powerful.

When I use Eclipse plugin (umlet 14.2), I cannot edit properties (text box in lower right corner). Also, my keyboard shortuts don't work (Undo - Command + Z, Paste etc.) but menu actions do.

More info:
`Eclipse IDE for Java Developers

Version: Neon Release (4.6.0)
Build id: 20160613-1800`

`OS X El Capitan 10.11.5`

`java version "1.8.0_73"
Java(TM) SE Runtime Environment (build 1.8.0_73-b02)
Java HotSpot(TM) 64-Bit Server VM (build 25.73-b02, mixed mode)`

**Comment by afdia** [comment]
Unfortunately this is an OSX specific problem (see #239) and we do not have a developer working on that OS to check those issues.

**Comment by zaster** [comment]
I can't find the menu when I put the plugin jar in eclipse/plugins

Eclipse IDE for java EE ,Version: Neon,System:Windows 10

**Comment by afdia** [comment]
You will only see the menu while you have a UMLet diagram opened.

**Comment by MarqueIV** [comment]
Same issue here.

macOS Sierra 10.12.6
Eclipse Photon (June 2018)
UMLet 14.3

Properties are read-only.  Also, UMLet keeps causing Eclipse to hang.  I have to force-quit Eclipse if I try interacting with the (read-only) property window.  Really frustrating.

**Comment by afdia** [comment]
Hi
This typically happens if the umlet plugin crashes. As mentioned above unfortunately it's OSX specific and  therefore hard to fix for me, but please check if there are any errors mentioned in the "Error Log" View (https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.user%2Freference%2Fref-error_log_view.htm)
You can also check the logs in "Help > About Eclipse > Installation detaul > Configuration" (or the file **"eclipse workspace"\.metadata\.log**) for information about the umlet plugin

If there is some information about the cause of the problem it would help us to fix it, but for the moment I would suggest to use standalone umlet or www.umletino.com which work without any issue on OSX.

**Comment by MarqueIV** [comment]
Ok, nothing interesting in the logs.  The only two entries I have are info about our proxy, and a message about BuiltinLFS, which apparently is known.

```
!ENTRY org.eclipse.core.net 1 0 2018-08-24 10:46:31.996
!MESSAGE System property http.nonProxyHosts has been set to [[redacted]] by an external source. This value will be overwritten using the values from the preferences

!ENTRY org.eclipse.egit.core 2 0 2018-08-24 10:46:32.039
!MESSAGE Builtin LFS support not present/detected
!STACK 0
java.lang.ClassNotFoundException: org.eclipse.jgit.lfs.BuiltinLFS cannot be found by org.eclipse.egit.core_5.0.0.201806131550-r
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:508)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:419)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:411)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:150)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        etc, etc...
```

I'm wondering if we're not seeing anything in the log because as soon as you click on the Properties window, it hangs Eclipse so perhaps nothing *can* get logged as it's' hung/crashed.

**Comment by afdia** [comment]
It sounds like the umlet plugin worked in previous versions. Is that correct and do you know when the problems started to appear? Perhaps we could narrow down the time the issue first occurred.
Of course it could also be a change in Eclipse/Java/... which causes the plugin to fail, then it would be harder to track down.

Is this issue specific to the properties panel? Does that mean you can open a diagram, move elements around, save and close it, open the palette dropdown and choose another one, add elements from the palette to the diagram, use the menu to e.g. zoom or export a diagram? Or in short: does it crash on any user interaction or only on specific actions?
We use a 3rd party lib to enhance the properties panel, so perhaps this causes problems if those other actions work.

Just a note about umletino: As you already mentioned, it doesn't send any diagrams anywhere (only if you use the optional dropbox integration) and you could even open the website from your local disk and it would still work (basically we compile Java to Javascipt and execute it), but I do understand that it may be problematic with the company policy and that not the whole functionality of the Java version is available.

About Umlet standalone: what do you mean with signing the application? Are you talking about signed jar files (https://docs.oracle.com/javase/tutorial/deployment/jar/signindex.html)? We use some dependencies which are probably not signed, so I guess only signing ours wouldn't help.

**Comment by MarqueIV** [comment]
Well, I can't speak to prior versions as I only found out about your project this week.

That said, here's what I've found.

(Assuming I don't touch the properties window...) I can always add new symbols, move around existing symbols, delete symbols (but not with the keyboard. I have to use the context menu or main menu) and it all works fine.  I can even *see* the properties for them just fine.  I can close/save and reopen them without issue.

Here's the odd thing.  If I just open a document but don't move or resize any elements, I can edit values in the property window without issue. However, once I have moved/resized an element on the canvas *after* editing the property window, the next time I go back to the property window, Eclipse hangs.

I've tried to see if I first select an element, then wait a while thinking maybe it has to do with the live compiling, but that doesn't seem to have an effect.  There's definitely something timing-wise going on though because almost always, if I open a diagram, click a node and immediately click in the properties window, that causes the hang too.

The only thing that's 100% certain is if I don't touch the properties window, everything works fine.  The minute I touch it, sometimes I can edit, sometimes it causes the hang.

Lemme know if you want me to do any specific tests based on the above info.

**Comment by MarqueIV** [comment]
Unrelated, how hard would it be for us to host our own copy of UMLetino?  Is that open-source too? (I confess, I haven't looked at that in much detail as the minute I saw 'Web site' I walked away.)

**Comment by afdia** [comment]
Thanks for the analysis, so it has probably something to do with the properties panel. We use a custom panel for things like autocompletion (since v12.0, which was some years ago), but I don't know if that is the cause.

You could try out v13.2 (https://www.umlet.com/umlet_13_2/com.umlet.plugin_13.2.0.jar) which should work according to the comment in the other issue.
Older versions are not provided via update site, therefore to install it, you must first uninstall v14.3, download the 13.2 jar file and put it into your Eclipse/dropins directory, then use "eclipse -clean" to make sure it reloads the plugin (sometimes downgrading to older versions of the plugin doesn't work and -clean tells eclipse to remove any plugin caches)
More infos are under https://github.com/umlet/umlet/wiki/Common-Problems#i-cannot-setup-the-eclipse-plugin

About umletino: As mentioned it lacks some features like lasso selection, "classic custom elements" and undo/redo of diagram movement (only text undo/redo works), but for the most part it does the same as the standalone/plugin version.
It is also open source and just a simple javascript application, therefore you can open it from your harddrive or host it yourself by putting the files on any http server.
We do not provide a downloadable package, but you can check out this git repository and build the projects using maven. You can check the travis build log to see how it should look: https://travis-ci.org/umlet/umlet
Basically you need Git (for the checkout) and Java+Maven for the build. If everything worked, you will see a zip file in the umlet-gwt/target directory which contains the full javascript based version of umlet

**Comment by MarqueIV** [comment]
Ok, first, I feel like an idiot! I fought with your instructions for a good thirty minutes with UMLet never loading the 13.2 version.  Tried the plugins directory in the app folder. Tried the plugins folder in the .p2/pool folder.  Tried a bunch of other locations.

...until I re-read your post for the fifth time and realized I had been skipping over *dropins* not plugins!

So the good news is that did load the 13.2 plugin.  The bad news is I'm still seeing the exact same behavior with it. When you first open a drawing, before you click on anything, the property window seems to work. But clicking on an object, then in the properties window, it hangs, just as before.

In the meantime I'm going to try and set up the UMLetino internally here.  Thanks for the link for that.  Still, management prefers the plugin (for some reason!) so I do want to still work on this.  My offer also still stands to share my screen or do whatever I can to help.  Lemme know.

**Comment by MarqueIV** [comment]
Also, did you include the UMLetino git repo?  You say 'this repo' but there's no link.  Just the ref to the regular UMLet code.) ...or are you saying UMLetino *is* in the source for UMLet 14.3?

**Comment by afdia** [comment]
Umletino shares ~60% of the code with the other platforms (by using GWT and compiling from Java to Javascript)
Also the Standalone and Eclipse Plugin are almost identical except for the initialization (once as a swing app, once embedded in Eclipse
Our wiki contains more information about the project structure if you're interested: https://github.com/umlet/umlet/wiki/Information-for-Developers

About the plugin: I hoped v13.2 would work because the author of the other issue mentioned it, but there were no changes which would explain the problem.
I have researched a bit and the first version which uses the current properties text area is v12.0, so if you have the time you could try out v11.5.1 (which was the last version with the old text area). As usual it's available on www.umlet.com (https://www.umlet.com/umlet_11_5_1/com.umlet.plugin_11.5.1.jar)
If 11.5.1 also has problems, I guess the issue is probably in the swing integration

Thanks for your effort!

**Comment by MarqueIV** [comment]
Ok, looks like I can confirm the issue is with BetterTouchTool, not UMLet, Eclipse or Java itself.  After uninstalling an reinstalling Eclipse (I think I screwed around with the files too much researching this so I wanted to start clean) and reinstalling UMLet 14.3 with BetterTouchTool closed, I could use all features of the plugin without issue. Moving, editing, saving, everything.

With Eclipse still open, but the UMLet editor closed, I re-launched BetterTouchTool and reopened the UMLet editor and sure enough.... hang!

Force-closed Eclipse, closed BetterTouchTool, relaunched Eclipse and reopened the diagram and it worked perfectly.  Reenabled BetterTouchTool and it hung again.

Now BetterTouchTool has app-specific configurations possible, so I added a rule for Eclipse to completely disable BetterTouchTool integration just for Eclipse, but to leave it on for the rest of the system.  Relaunched the UMLet editor, and it works 100%.

I think we found the culprit.  You may want to add a note in your FAQ about this.  Simply say 'There are known issues with BetterTouchTool on the Mac.  To get UMLet to work properly, in the preferences for BetterTouchTool, add an application-specific configuration for Eclipse and set it to completely disable BetterTouchTool for that application, and you should be able to work with it." ...or something to that effect.

Let me know if you want me to provide screenshots and such that you can add. 

In the meantime, I'm actually going to reach out to the BetterTouchTool dev and let him know there's an issue with his app and UMLet, but it may be with a package you're using.  Either way, we now have 100% reproducible and consistent behaviors, and more importantly, a work-around.

So glad I got to the bottom of this, and happier that it doesn't require you to do anything from your code's side. :)

**Comment by MarqueIV** [comment]
Just for reference, I'm attaching that log here...

[EclipseCrash.txt](https://github.com/umlet/umlet/files/2324967/EclipseCrash.txt)

**Comment by afdia** [comment]
Wow thanks that sounds great!
I will try to find other mac osx users to verify the behavior, but I'm very happy that you have found the cause (or let's say the problematic combination of applications).
Now we can provide a workaround until it will hopefully be fixed in the future.

Perhaps you could still try v11.5.1 to check my theory that the switch from jsyntaxpane to rsyntaxarea (both enhanced java text areas) introduced the issue, but even if not thanks for your help!

**Comment by MarqueIV** [comment]
So you can follow along as well, I've opened a bug report with him...

https://community.folivora.ai/t/bettertouchtool-causes-eclipse-to-hang-when-using-the-umlet-plug-in/3269

Lemme see if I can test your theory about that older version. More to come...

M

**Comment by MarqueIV** [comment]
Well, just uninstalled 14.3 via the market, copied your 11.5.1 to the 'dropins' folder, but whenever I try launching UMLet, it hangs even with BetterTouchTool closed, so I can't test that theory. Sorry about that.  I'll let you know if I get anywhere else.

**Comment by MarqueIV** [comment]
Ok, same thing's happening to me now that happened before.  Once I uninstall UMLet, try an older version via the 'dropins', even if I remove the file from 'dropins', run 'eclipse -clean' before reinstalling UMLet, install UMLet, then run 'eclipse -clean' again, even though it tells me UMLet is installed, it doesn't show up anywhere and opening a file only shows the XML.  Not sure if it's because UMLet was run from two different places. Gonna try nuking my workspace and try again.  If that doesn't work, I'm going to have to uninstall and reinstall Eclipse again, like I did last time.

**Comment by fifafu** [comment]
BetterTouchTool dev here. 
The crashlogs shows that this is happening when accessing the currently focused UI element through the macOS accessibility API. This should be supported by all apps, but possibly Eclipse doesn't.

However BTT should not access that unless you are using the new "Conditional Activation Group" feature. As this is a **very** new feature - are you using it? Or do I have a bug somwhere that causes this to be enabled?

if you are using a conditional activation group, can you check whether this option in BTT fixes the issue?
![image](https://user-images.githubusercontent.com/8820811/44680246-735cf900-aa3d-11e8-92fd-d6cd3070d4a9.png)

**Comment by MarqueIV** [comment]
Hi Andreas!  You and I have actually spoken several times over the years.  Huge fan of your work. Long-time paid supporter. I even tried to get you to partner with my company Inertial Design for those keyboard/trackpad aluminum trays years ago.

Anyway, thanks for getting back so quickly.  I can't speak to what the dev is doing because I'm just a user of UMLet, and the dev doesn't own a Mac (which makes this really fun to try to debug. It's like one person driving blindfold while the passenger tells them what to do.) Next time I'm on that computer (tomorrow morning) I'll give that a try to see if that option works for Eclipse.

I can say this though... the dev mentioned he switched to using a different text area package back in version 12 so if that does turn out to be the culprit, maybe we need to get those devs involved too.

Just out of curiosity, what exactly does that do/disable?  Does that affect my ability to send custom keystrokes, etc.?  What about window resizing.  Right now, I have 'Disable BTT completely' so I lose everything when in there. Just wondering what I'll get back with this, especially if it works.

Thanks for your quick reply, and keep up the awesome work!

M

**Comment by MarqueIV** [comment]
Hi Andreas! You and I have actually spoken several times over the years. Huge fan of your work. Long-time paid supporter. I even tried to get you to partner with my company Inertial Design for those keyboard/trackpad aluminum trays years ago.

Anyway, thanks for getting back so quickly. To answer your question, no, I'm not using the 'Conditional Activation Group' feature. Not even sure what it is! I actually haven't really touched my config in years, save adding a gesture/trigger or two. It's pretty damn perfect the way it is so I tend not to touch it too much.

As for your suggestion, tomorrow morning when I'm back on that machine, I'll give that a try to see if that option works for Eclipse.

As for anything with the UMLet code itself, I can't speak to what the dev is doing because I'm just a user of UMLet, and the dev doesn't own a Mac (which makes this really fun to try to debug. It's like one person driving blindfold while the passenger tells them what to do.)

I can say this though... Eclipse itself seems to work fine with BTT. It's only that plugin I've had trouble with. The dev mentioned he switched to using a different text area package back in version 12, and interacting with the text area is what causes the hang, so if that turns out to be the culprit, maybe we need to get the package devs involved too. (That I'll have to defer to the UMLet dev to get more info on though.)

Just out of curiosity, as for the setting you want me to try, what exactly does that do/disable? Does that affect my ability to send custom keystrokes, etc.? What about window resizing. Right now, I have 'Disable BTT completely' so I lose everything when in there. Just wondering what I'll get back with this, especially if it works.

Thanks for your quick reply, and keep up the awesome work!

M

**Comment by MarqueIV** [comment]
On a side-note, it's pretty funny to see the developer of a program having to run a trial version of his own work! You should pay yourself, mate!  You do good work! ;)

**Comment by MarqueIV** [comment]
Ha! I just realized both of you are named Andreas! Guess they give you that name when you write cool software. ;)

Well, "UMLet Andreas", this part's for you.  Just ran your latest plugin on the macOS Mojave beta and I'm seeing the same thing as another user from a different issue... I can see the properties window... I can click in the properties window... I can even select things in the properties window, but I can't *type* in the properties window. It's read-only.  Again, this isn't a hang, and BTT is not running at all in this case.  Just another $#% issue with that property box.

Note, the stand-alone UMLet seems to work fine on the Mojave beta. It's only the Eclipse plug-in.

**Comment by afdia** [comment]
I guess it's a common name :)
Thanks for the additional testing, so to summarize:
* v11.5.1 doesn't work at all at machine 1 (macOS Sierra)
* v14.3 has problems on machine 1 (macOS Sierra) which can be fixed by disabling BTT and slightly different problems on machine 2 (macOS Mojave) which isn't running BTT?

Have you tried 11.5.1 on macOS Mojave?
I could try to make a custom build with a default textarea instead to check my theory or perhaps try to improve exception loggin (as described in https://wiki.eclipse.org/Eclipse_Plug-in_Development_FAQ#How_do_I_show_a_message_dialogue_for_exceptions_and_log_them.3F )

**Comment by afdia** [comment]
I've added some code which should log uncaught exceptions (I always assumed that they are logged already)
Here is a snapshot build to try out: https://www.dropbox.com/s/i2sp87y48058itd/umlet-eclipse-p2-14.4.0-SNAPSHOT.zip?dl=1
It's a p2 repository, so just download it and unzip it.
In Eclipse use "Help > Install New Software > Add > Local", locate the unzipped "repository" file and install "UMLet Eclipse Plugin"

If you prefer to build it yourself, you can also check out the umlet repository, do a maven build (as described above) and use the zip file which is located at umlet/umlet-eclipse-p2/target/umlet-eclipse-p2-14.4.0-SNAPSHOT.zip

**Comment by MarqueIV** [comment]
I'm a bit slammed today (and have a New York Red Bulls soccer match to go to tonite #NYCIsRed!!!) but I'll either check this tonight after the game, or tomorrow when I'm working from home.

Unrelated, I found an unbelievably simple way to get around the 'Unknown Developer' issue on our Macs here for the stand-alone version.  It's kinda embarrassing how you get around it.

1. Download and unzip standalone UMLet as normal
2. Open the folder, right click on and unzip 'UMLet.jar' (which is really just a zipped archive) into a temp directory.
3. Go into the temp directory, select everything, and re-zip it up, naming the resulting file 'UMLet.jar'
4. Replace the original 'UMLet.jar' file (back in the original directory) with the one you just created.
5. (For completeness) delete the temp folder (where you extracted 'UMLet.jar' into)
6. Enjoy your newly-trusted, stand-alone version of the app!

Not sure why, but the above steps (found in a YouTube video!) suddenly makes the app from a known developer (me, I guess?) and then it runs just fine! No Eclipse needed!

Still, I'll be more than willing to work with you on figuring out the Eclipse issues, but at least I'm now running your app in stand-alone, which is what I wanted to do in the first place.

**Comment by MarqueIV** [comment]
I'm a bit slammed today (and have a New York Red Bulls soccer match to go to tonite #NYCIsRed!! WOOT!!!) but I'll either check this tonight after the game, or tomorrow when I'm working from home.

Tangentially related, I found an unbelievably simple way to get around the 'Unknown Developer Blocked' issue on our Macs here thanks to a YouTube video. It's kinda embarrassing how easy it is to do considering I literally could teach my mom how to do it (and she once thought she broke the entire internet! True story!)

Download and unzip standalone UMLet as normal
Open the folder, right click on and unzip 'UMLet.jar' (which is really just a zipped archive) into a temp directory.
Go into the temp directory, select everything, and re-zip it up, naming the resulting archive file 'UMLet.jar'
Replace the original 'UMLet.jar' (back in the original directory) with the one you just created.
(For completeness) delete the temp folder
Enjoy your new life as a master, Black Hat god, thriving in the deepest corners of the dark web, laughing at the systems around you that you can now PWN!!
6b. (Alternate ending) Laugh at how stupidly simple that was to get around.
So yep, I'm now running stand-alone without issue. Still, I'll be more than willing to work with you on figuring out the Eclipse issues, but at least I'm now running your app in stand-alone, which is what I wanted to do in the first place.

**Comment by MarqueIV** [comment]
I'm a bit slammed today (and have a New York Red Bulls soccer match to go to tonite #NYCIsRed!! WOOT!!!) but I'll either check this tonight after the game, or tomorrow when I'm working from home.

Tangentially related, I found an unbelievably simple way to get around the 'Unknown Developer Blocked' issue on our Macs here thanks to a YouTube video. It's kinda embarrassing how easy it is to do considering I literally could teach my mom how to do it (and she once thought she broke the entire internet! True story!)

1. Download and unzip standalone UMLet as normal
2. Open the folder, right click on and unzip 'UMLet.jar' (which is really just a zipped archive) into a temp directory.
3. Go into the temp directory, select everything, and re-zip it up, naming the resulting archive file 'UMLet.jar'
4. Replace the original 'UMLet.jar' (back in the original directory) with the one you just created.
5. (For completeness) delete the temp folder
6. Enjoy your new life as a master, Black Hat god, thriving in the deepest corners of the dark web, laughing at the systems around you that you can now PWN!!
6b. (Alternate) Laugh at how stupidly simple that was to get around.

So yep, I'm now running stand-alone without issue. Still, I'll be more than willing to work with you on figuring out the Eclipse issues, but at least I'm now running your app in stand-alone, which is what I wanted to do in the first place.

**Comment by afdia** [comment]
Good to hear that you can finally use standalone UMLet, as for helping out with the plugin please check if you see additional logging with the provided 14.4 snapshot build.

Other than that I'm not sure ... it's interesting that there seems to be a connection to BetterTouchTool but I'm not sure if there is anything we could fix or if it's just a weird coincidence that the plugin stops crashing after disabling BetterTouchTool

**Comment by MarqueIV** [comment]
Few things.

To confirm your earlier question:

Yes, the eclipse crash/hang seems to be caused by a combination of what you're currently using in UMLet, and how BTT interacts with the system via accessibility.

Yes, that is a separate issue from when UMLet's properties window becomes read-only.  I believe someone said that was because the (3rd-party) code behind the properties window crashes, thus turning it into a read-only window since all interactions are disabled.  It does seem to still *display* the correct data. You just can't interact with it.

Also, kinda slapping my forehead on this, but there's an even easier way to allow UMLet to run on machines that block unknown developers.  You download and unzip it as usual, then execute this line from terminal:

`latter -d com.apple.quarantine path/to/UMLet.jar`

That's it.  It now opens right up. No need for all that unzipping/re-zipping nonsense (and now even more laughably simple! You don't even need sudo!)

Anyway, as promised, give me a little time and I'll try all the variants of Eclipse/macOS/BTT with your new snapshot build and will get back to you.

HTH,

M
